semgrep-core/src/tainting/Dataflow_tainting.ml

```diff
@@ -383,128 +412,171 @@ let handle_taint_propagators env x taints =
+       * annotation, then we infer that the variable itself is now tainted
+       * (presumably by side-effect) and we will update the `var_env`
+       * accordingly. Otherwise the variable belongs to a piece of code that
+       * is a source of taint, but it is not tainted on its own. *)
+    List.partition is_exact source_pms
   in
-  match sanitizer_pms with
-  | _ :: _ -> (Taints.empty, env.lval_env)
-  | [] ->
-      let taints = source_pms |> taints_of_matches in
-      let sinks = sink_pms |> Common.map trace_of_match in
-      let findings = findings_of_tainted_sinks env taints sinks in
-      report_findings env findings;
-      (taints, env.lval_env)
-
-(* Test whether a variable occurrence is tainted, and if it is also a sink,
- * report the finding too (by side effect).
- *
- * THINK: Could we generalize this for lvalues? So we won't need special
- *     Lval_env.add_var and Lval_env.find_var, but just work with lvalues
- *     everywhere?
- *)
-let check_tainted_var env (var : IL.name) : Taints.t * Lval_env.t =
-  let source_pms, sanitizer_pms, sink_pms =
-    let _, tok = var.ident in
-    if Parse_info.is_origintok tok then
-      ( env.config.is_source (G.Tk tok),
-        env.config.is_sanitizer (G.Tk tok),
-        env.config.is_sink (G.Tk tok) )
-    else ([], [], [])
+  let taints_sources_reg =
+    reg_source_pms |> taints_of_matches |> filter_taints_by_labels labels
+  and taints_sources_mut =
+    mut_source_pms |> taints_of_matches |> filter_taints_by_labels labels
   in
-  match sanitizer_pms with
+  let lval_env = Lval_env.add lval taints_sources_mut env.lval_env in
+  (Taints.union taints_sources_reg taints_sources_mut, lval_env)
+
+let rec check_tainted_lval env (lval : IL.lval) : Taints.t * Lval_env.t =
+  let new_taints, lval_in_env, lval_env = check_tainted_lval_aux env lval in
+  let taints_from_env = status_to_taints lval_in_env in
+  let taints = Taints.union new_taints taints_from_env in
+  let sinks = lval_is_sink env.config lval |> Common.map trace_of_match in
+  let findings = findings_of_tainted_sinks { env with lval_env } taints sinks in
+  report_findings { env with lval_env } findings;
+  (taints, lval_env)
+
+and check_tainted_lval_aux env (lval : IL.lval) :
+    Taints.t * [ `Clean | `None | `Tainted of Taints.t ] * Lval_env.t =
+  (* Recursively checks an l-value bottom-up.
+   *
+   *  This check needs to combine matches from pattern-{sources,sanitizers,sinks}
+   *  with the info we have stored in `env.lval_env`. This can be subtle, see
+   *  comments below.
+   *)
+  match lval_is_sanitized env.config lval with
   (* TODO: We should check that taint and sanitizer(s) are unifiable. *)
-  | _ :: _ ->
-      let lval_env' = sanitize_var env.lval_env sanitizer_pms var in
-      (Taints.empty, lval_env')
+  | _ :: _ as sanitizer_pms ->
+      (* NOTE [lval/clean]:
+       *  If lval is sanitized, then we will "bubble up" the `Clean status, so any
```

Comment:
I think it goes "up" ? You can probably find a better name than me. But essentially if we have `x.a.b.c`, and we're checking `x.a` and it's "clean" (e.g. because it has been sanitized), then we "propagate up" the "clean" to `x.a.b` and `x.a.b.c`.